aea/manager.py

```diff
@@ -0,0 +1,534 @@
+
+        self._error_callbacks: List[Callable[[str, BaseException], None]] = []
+
+        if mode not in self.MODES:
+            raise ValueError(
+                f'Invalid mode {mode}. Valid modes are {", ".join(self.MODES)}'
+            )
+        self._started_event = threading.Event()
+        self._mode = mode
+
+    @property
+    def is_running(self) -> bool:
+        """Is manager running."""
+        return self._is_running
+
+    def _run_thread(self) -> None:
+        """Run internal thread with own event loop."""
+        self._loop = asyncio.new_event_loop()
+        self._event = asyncio.Event(loop=self._loop)
+        self._loop.run_until_complete(self._manager_loop())
+
+    async def _manager_loop(self) -> None:
+        """Perform manager stop."""
+        if not self._event:
+            raise ValueError("Do not use this method directly, use start_manager.")
+
+        self._started_event.set()
+
+        while self._is_running:
+            tasks_for_agents = {
+                task.wait(): agent_name
+                for agent_name, task in self._agents_tasks.items()
+            }
+            wait_tasks = list(tasks_for_agents.keys()) + [self._event.wait()]  # type: ignore
+            done, _ = await asyncio.wait(wait_tasks, return_when=FIRST_COMPLETED)
+
+            if self._event.is_set():
+                self._event.clear()
+
+            for task in done:
+                if task not in tasks_for_agents:
+                    await task
+                    continue
+                agent_name = tasks_for_agents[task]
+                self._agents_tasks.pop(agent_name)
+                if task.exception():
+                    print(agent_name, "exceptioned", task.exception())
+                    for callback in self._error_callbacks:
+                        callback(agent_name, task.exception())
+                else:
+                    await task
+
+    def start_manager(self) -> None:
+        """Start manager."""
+        self._ensure_working_dir()
+        self._started_event.clear()
+        self._is_running = True
+        self._thread.start()
+        self._started_event.wait(self.DEFALUT_TIMEOUT_FOR_BLOCKING_OPERATIONS)
+
+    def stop_manager(self, cleanup: bool = False) -> None:
+        """
+        Stop manager.
+
+        Stops all running agents and stop agent.
+
+        :param cleanup: remove agents_dir and purge in memory registry.
+
+        :return: None
+        """
+        if not self._loop or not self._event:
+            raise ValueError("Manager was not started!")
+        if self._was_working_dir_created:
+            rmtree(self.working_dir)
+
+        if self._thread.is_alive():
+            self.stop_all_agents()
+        self._is_running = False
+        self._loop.call_soon_threadsafe(self._event.set)
+        self._thread.join(self.DEFALUT_TIMEOUT_FOR_BLOCKING_OPERATIONS)
+
+    def add_project(self, public_id: PublicId) -> None:
```

Comment:
We can use the same pattern as in the AEA builder and return the manager; that way you can do chained manager operations like `manager.add_project(id_1).add_project(id_2)` etc.